certbot/compat/filesystem.py

```diff
@@ -0,0 +1,48 @@
+"""Compat module to handle files security on Windows and Linux"""
+from __future__ import absolute_import
+
+import os  # pylint: disable=os-module-forbidden
+
+from acme.magic_typing import Callable  # pylint: disable=unused-import,no-name-in-module
+
+
+def makedirs(file_path, mode=0o777):  # pylint: disable=function-redefined
+    # type: (str, int) -> None
+    """
+    Wrapper of original os.makedirs function, that will ensure on Windows that given mode
+    is correctly applied.
+    :param str file_path: The file path to open
+    :param int mode: POSIX mode to apply on file when opened,
+        Python defaults will be applied if ``None``
+    """
+    # As we know that os.mkdir is called internally by os.makedirs, we will swap the function in
+    # os module for the time of makedirs execution.
+    orig_mkdir_fn = os.mkdir
+    try:
+        def wrapper(one_path, one_mode=0o777):  # pylint: disable=missing-docstring
+            # Note, we need to provide the origin os.mkdir to our mkdir function,
+            # or we will have a nice infinite loop ...
+            mkdir(one_path, mode=one_mode, mkdir_fn=orig_mkdir_fn)
+
+        os.mkdir = wrapper
+
+        os.makedirs(file_path, mode)
```

Comment:
I don't love that we're relying on the internal details of `os.makedirs`' implementation for this to work. The implementation has to only be a few lines of recursive code that we could easy reuse here, right?

Reply:
I agree that I could reimplement the makedirs logic here, but it is not the design I want to follow.

The idea here is to make filesystem module as close as possible to the original os module capabilities. I do that by following a design consisting in reusing at most the existing logic in os, and change only what needs to be changed to apply the security model. 

Here the documentation says that makedirs use mkdir internally, so it is a documented entrypoint that I think suitable to use. However, I do not want to make any other assumption on how makedirs work for the several python versions that we can encounter. I prefer to consider most of the logic of makedirs as a black box, and stay consistent to the actual behavior of it in a given python version, instead of fixing a global unique logic for all of them.

Reply:
I will add that the specific case that I want to cover with this design, is that if for any reason python maintainers wants to change some behavior in makedirs, I want to take advantage of it flawlessly.